pop-api/examples/fungibles/lib.rs

```diff
@@ -1,108 +1,268 @@
+		///
+		/// # Parameters
+		/// - `spender` - The account that is allowed to spend the tokens.
+		/// - `value` - The number of tokens to increase the allowance by.
 		#[ink(message)]
-		pub fn increase_allowance(
+		fn increase_allowance(
 			&mut self,
-			token: TokenId,
 			spender: AccountId,
 			value: Balance,
-		) -> Result<()> {
-			api::increase_allowance(token, spender, value)
+		) -> Result<(), Psp22Error> {
+			let caller = self.env().caller();
+			// No-op if the caller and `spender` is the same address or `value` is zero.
+			if caller == spender || value == 0 {
+				return Ok(());
+			}
+			api::increase_allowance(self.id, spender, value).map_err(Psp22Error::from)?;
+			let allowance = self.allowance(caller, spender);
+			self.env().emit_event(Approval { owner: caller, spender, value: allowance });
+			Ok(())
 		}
 
+		/// Decreases the allowance of `spender` by `value` amount of tokens.
+		///
+		/// # Parameters
+		/// - `spender` - The account that is allowed to spend the tokens.
+		/// - `value` - The number of tokens to decrease the allowance by.
 		#[ink(message)]
-		pub fn decrease_allowance(
+		fn decrease_allowance(
 			&mut self,
-			token: TokenId,
 			spender: AccountId,
 			value: Balance,
-		) -> Result<()> {
-			api::decrease_allowance(token, spender, value)
+		) -> Result<(), Psp22Error> {
+			let caller = self.env().caller();
+			// No-op if the caller and `spender` is the same address or `value` is zero.
+			if caller == spender || value == 0 {
+				return Ok(());
+			}
+			api::decrease_allowance(self.id, spender, value).map_err(Psp22Error::from)?;
+			let value = self.allowance(caller, spender);
+			self.env().emit_event(Approval { owner: caller, spender, value });
+			Ok(())
+		}
+	}
+
+	impl Psp22Metadata for Fungible {
+		/// Returns the token name.
+		#[ink(message)]
+		fn token_name(&self) -> Option<String> {
+			api::token_name(self.id)
+				.unwrap_or_default()
+				.and_then(|v| String::from_utf8(v).ok())
 		}
 
+		/// Returns the token symbol.
 		#[ink(message)]
-		pub fn token_name(&self, token: TokenId) -> Result<Vec<u8>> {
-			api::token_name(token)
+		fn token_symbol(&self) -> Option<String> {
+			api::token_symbol(self.id)
+				.unwrap_or_default()
+				.and_then(|v| String::from_utf8(v).ok())
 		}
 
+		/// Returns the token decimals.
 		#[ink(message)]
-		pub fn token_symbol(&self, token: TokenId) -> Result<Vec<u8>> {
-			api::token_symbol(token)
+		fn token_decimals(&self) -> u8 {
+			api::token_decimals(self.id).unwrap_or_default()
 		}
+	}
+
+	impl Psp22Mintable for Fungible {
+		/// Creates `value` amount of tokens and assigns them to `account`, increasing the total
+		/// supply.
+		///
+		/// # Parameters
+		/// - `account` - The account to be credited with the created tokens.
+		/// - `value` - The number of tokens to mint.
+		#[ink(message)]
+		fn mint(&mut self, account: AccountId, value: Balance) -> Result<(), Psp22Error> {
```

Comment:
Surely this should be protected? Anyone can call this contract and mint Balance::MAX.

This example should be improved to have some origin set at instantiation which checks for minting/burning. See https://docs.openzeppelin.com/contracts/2.x/api/token/erc20#ERC20Mintable